Add WebSocketStream article (#2441)
GoogleChrome/web.dev: src/site/content/en/blog/websocketstream/index.md

---

> title: WebSocketStream—Integrating Streams with the WebSocket API

Reviewer: ```suggestion
title: WebSocketStream: integrating streams with the WebSocket API
```

Author: It needs to be on a new line due to the colon, but yes.

subhead: |
  Prevent your app from getting drowned in WebSocket messages
  or flooding a WebSocket server with messages by applying backpressure.
authors:
  - thomassteiner
date: 2020-03-24

> update: 2020-03-24

Reviewer: ```suggestion
date: 2020-03-24
```

Author: This is a moving target :-)

hero: hero.jpg
alt: A fire hose with water dripping out of it.
description: |
  WebSocketStream integrates streams with the WebSocket API.
  This allows your app to apply backpressure to received messages.
tags:
  - post # post is a required tag for the article to show up in the blog.
  - backpressure
  - websocket
  - websocketstream
  - capabilities
  - fugu
draft: true
origin_trial:
  url: https://developers.chrome.com/origintrials/#/view_trial/1977080236415647745
---
## Background

### The WebSocket API

The [WebSocket API](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API)
provides a JavaScript interface to the [WebSocket protocol](https://tools.ietf.org/html/rfc6455),
which makes it possible to open a two-way interactive communication session
between the user's browser and a server.
With this API, you can send messages to a server and receive event-driven responses

> without having to poll the server for a reply.

Reviewer: ```suggestion
without polling the server for a reply.
```

### The Streams API

The [Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API)
allows JavaScript to programmatically access streams of data chunks received over the network

> and process them as desired by the developer.

Reviewer: ```suggestion
and process them as desired.
```

An important concept in the context of streams is
[backpressure](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Concepts#Backpressure).
This is the process by which a single stream or a pipe chain
regulates the speed of reading or writing.
When the stream itself or a stream later in the pipe chain is still busy
and isn't yet ready to accept more chunks,

> it sends a signal backwards through the chain to slow down delivery as appropriate.

Reviewer: ```suggestion
it sends a signal backwards through the chain to slow delivery as appropriate.
```

### The Problem with the current WebSocket API

#### Applying backpressure to received messages is impossible

With the current WebSocket API, reacting to a message happens via the
[`WebSocket.onmessage`](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/onmessage)

> property, an `EventHandler` that is called when a message is received from the server.

Reviewer: ```suggestion
With the current WebSocket API, reacting to a message happens in
[`WebSocket.onmessage`](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/onmessage)
, an `EventHandler` called when a message is received from the server.
```

> Let's assume you had an application that needed to perform some heavy data crunching operations

Reviewer: ```suggestion
Let's assume you had an application that needs to perform heavy data crunching operations
```

whenever a new message is received.

> You would probably set up the flow similar to the code in the example below,

Reviewer: ```suggestion
You would probably set up the flow similar to the code below,
```

    }, 1000);
  });
};

webSocket.onmessage = async (event) => {
  const data = event.data;
  // Await the result of the processing step in the message handler.
  await process(data);
};
```

Wrong! The problem with the current WebSocket API is that there is no way to apply backpressure.
When messages arrive faster than the `process()` method can handle them,
the render process will either fill up memory by buffering those messages,
become unresponsive due to 100% CPU usage, or both.

#### Applying backpressure to sent messages is non-ergonomic

Applying backpressure to sent messages is possible, but involves polling the
[`WebSocket.bufferedAmount`](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/bufferedAmount)
property, which is inefficient and non-ergonomic.
This read-only property returns the number of bytes of data that have been queued
using calls to
[`WebSocket.send()`](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/send),
but not yet transmitted to the network.
This value resets to zero once all queued data has been sent,
but if you keep calling `WebSocket.send()`,
it will continue to climb.

## What is the WebSocketStream API? {: #what }

> The new WebSocketStream API deals with the problem of non-existent or non-ergonomic backpressure

Reviewer: ```suggestion
The WebSocketStream API deals with the problem of non-existent or non-ergonomic backpressure
```

Reviewer: Everyone who reads your article now, will probably know the API is new. It won't be new if someone reads this article two years from now.

> applicability by integrating Streams with the WebSocket API.

Reviewer: ```suggestion
applicability by integrating streams with the WebSocket API.
```

> applicability by integrating Streams with the WebSocket API.

Reviewer: Applicability? You can probably eliminate this word.

Author: I will :-)

> This means backpressure can be applied "for free", without any extra hoops.

Reviewer: ```suggestion
This means backpressure can be applied "for free", without any extra cost.
```

Reviewer: I don't usually quibble with a writers choice of idiom. Shortening 'jumping through hoops' to 'hoops' may cause problems for readers who are not fluent in English. Cost is easier to translate than a figure of speech is.

### Suggested use cases for the WebSocketStream API {: #use-cases }

> Examples of sites that may use this API include:

Reviewer: ```suggestion
Examples of sites that can use this API include:
```

* High-bandwidth WebSocket applications that need to retain interactivity,
  in particular video and screen-sharing.
* Similarly, video capture and other applications that generate a lot of data in the browser
  that needs to be uploaded to the server.

>   With backpressure, the client can stop producing data rather than accumulating data in memory.

Reviewer: ```suggestion

With backpressure, the client can stop producing data rather than accumulating data in memory.
```

Reviewer: Am I correct that this line was not meant to be part of the second bullet?

Author: It actually was part of the bullet.

| 3. Gather feedback & iterate on design     | [In progress](#feedback)     |
| 4. Origin trial                            | [Complete][ot]               |
| 5. Launch                                  | Not started                  |

</div>

## How to use the WebSocketStream API {: #use }

### Introductory example

The WebSocketStream API is promise-based, which makes dealing with it feel natural
in a modern JavaScript world.
You start by constructing a new `WebSocketStream` and passing it the URL of the WebSocket server.
Next, you wait for the `connection` to be established,
which results in a
[`ReadableStream`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/ReadableStream)
and/or a
[`WritableStream`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream/WritableStream).

By calling the
[`ReadableStream.getReader()`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/getReader)
method, you finally obtain a
[`ReadableStreamDefaultReader`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultReader),
which you can then [`read()`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultReader/read)

> data from until it is done, that is, until it returns an object of the form

Reviewer: ```suggestion
data from until it the stream done, that is, until it returns an object of the form
```

`{value: undefined, done: true}`.

Accordingly, by calling the
[`WritableStream.getWriter()`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream/getWriter)
method, you finally obtain a
[`WritableStreamDefaultWriter`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultWriter),
which you can then [`write()`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultWriter/write)
data to.

```js
  const readWSS = new WebSocketStream(READ_URL);
  const {readable} = await readWSS.connection;
  const reader = readable.getReader();

  const writeWSS = new WebSocketStream(WRITE_URL);
  const {writable} = await writeWSS.connection;
  const writer = writable.getWriter();

  while (true) {
    const {value, done} = await reader.read();
    if (done) {
      break;
    }
    const result = await process(value);
    await writer.write(result);
  }
```

#### Backpressure

What about the promised backpressure feature?
As I wrote above, you get it "for free", no extra steps needed.

> If `process()` takes a second, the next message will only be consumed once the pipeline is ready.

Reviewer: ```suggestion
If `process()` takes extra time, the next message will only be consumed once the pipeline is ready.
```

Reviewer: I didn't like 'a second' because it implies, ... well, exactly one second.

Likewise for the `WritableStreamDefaultWriter.write()` step:

> it will only proceed if it is safe to do so.

Reviewer: ```suggestion
Likewise the `WritableStreamDefaultWriter.write()` step
will only proceed if it is safe to do so.
```

### Advanced examples

The second argument to WebSocketStream is an option bag to allow for future extension.
Currently the only option is `protocols`,
which behaves the same as the
[second argument to the WebSocket constructor](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket#Parameters:~:text=respond.-,protocols):

```js
const chatWSS = new WebSocketStream(CHAT_URL, {protocols: ['chat', 'chatv2']});

> const {protocol} = await wss.connection;

Author: ```suggestion
const {protocol} = await chatWSS.connection;
```

> const {protocol} = await wss.connection;

Reviewer: Should `wss` be `chatWSS`?

Author: Good catch! Yes!

```

The selected protocol is part of the dictionary available
via the `WebSocketStream.connection` promise,

> along with `extensions`.

Author: ```suggestion
The selected `protocol` as well as potential `extensions` are part of the dictionary
available via the `WebSocketStream.connection` promise.
```

> along with `extensions`.

Reviewer: I don't understand this sentence.

Author: I rephrased.

All the information about the live connection is provided by this promise,

> since it is not relevant if the connection failed.

Reviewer: ```suggestion
since it is not relevant if the connection fails.
```
